src/coreComponents/finiteElement/kernelInterface/KernelBase.hpp

```diff
@@ -317,16 +292,158 @@ class KernelFactory
+                            LvArray::system::demangleType< CONSTITUTIVE_TYPE >() + ", " +
+                            LvArray::system::demangleType< FE_TYPE >() + ", " +
+                            string( NAME ) + ", " +
+                            LvArray::system::demangleType< decltype( m_args ) >();
+
+      // Unfortunately can't just decltype(&buildKernelAndInvoke) since we can't fully specify the function template
+      using JIT_KERNEL_DISPATCH = real64 (*)( localIndex const, 
+                                              NodeManager &,
+                                              EdgeManager const &,
+                                              FaceManager const &,
+                                              localIndex const, 
+                                              SUBREGION_TYPE const &, 
+                                              FE_TYPE const &, 
+                                              CONSTITUTIVE_TYPE &, 
+                                              decltype( m_args ) const & );
+      string outputDir( STRINGIZE( JITTI_OUTPUT_DIR ) );
+      outputDir += "/";
+      static jitti::Cache< JIT_KERNEL_DISPATCH > buildCache( time(NULL), outputDir );
+
+      if( MpiWrapper::commRank( ) == 0 )
+      {
+        buildCache.getOrLoadOrCompile( info );
+      }
+      MpiWrapper::barrier( );
+      // check if the library with the function is available
+      if ( buildCache.tryGet( info ) == nullptr )
+      {
+        // if not, refresh by reading the filesystem to find the new library on the first iteration
+        buildCache.refresh( );
+      }
+      auto & jitKernelDispatch = buildCache.getOrLoad( info );
+      return jitKernelDispatch( numElems,
+                                nodeManager, 
+                                edgeManager, 
+                                faceManager,
+                                targetRegionIndex,
+                                elementSubRegion,
+                                finiteElementSpace,
+                                inputConstitutiveType,
+                                m_args );
+    }
+
+  private:
+
+    /// The arguments to append to the standard kernel constructor arguments.
+    camp::tuple< ARGS ... > m_args;
+  };
 
+// Would VASTLY prefer to use template specialization on the above class(es) and allow JITTI_TPARAM to ultimately 
+//  decide on whether to JIT or not.. but current language restrictions on templating and our in-code restrictions on
+//   the kernels due to the 'using' statements throughout the code to define specific kernel dispatchers make that
+//   devwork intractable.
+// We can't use 'using' here due to differences in the above templates as well... nor a nice preprocessor macro... 
+//   so we're basically left with only one option
+#if JITTI == 1
+  #define KernelDispatch KernelDispatchJIT
+#else
+  #define KernelDispatch KernelDispatchTemplate
+#endif
```

Comment:
Per the comments, this is the part of the current implementation I want to change the most. It should be possible, but wound up eating hours while I was trying to get it implemented the way I would prefer.

Reply:
It should be possible to make the two dispatcher classes similar by replacing `NAME` and `HEADER` parameters in `KernelDispatchJIT` with with just `KERNEL_TYPE` parameter (i.e. template template parameter, like in the non-JIT dispatcher).

`NAME` could then be extracted as (I'm trying to decide if this is 100% robust):
```c++
string const fullName = LvArray::system::demangleType< KERNEL_TYPE< SUBREGION_TYPE, FE_TYPE, CONSTITUTIVE_TYPE > >();
string const name = fullName.substr( 0, fullName.find( '<' ) );
```

`HEADER` could be replaced by adding inside each "leaf" kernel class something like
```c++
template< typename SUBREGION_TYPE, typename FE_TYPE, typename CONSTITUTIVE_TYPE >
class MyKernel
{
  static constexpr char const source_location[] = __FILE__;
  ...
};

// Sad part: this is required in C++14 to avoid linker errors
template< typename SUBREGION_TYPE, typename FE_TYPE, typename CONSTITUTIVE_TYPE >
constexpr char const MyKernel< SUBREGION_TYPE, FE_TYPE, CONSTITUTIVE_TYPE >source_location[];
```
and accessing as
```c++
string const header = KERNEL_TYPE< SUBREGION_TYPE, FE_TYPE, CONSTITUTIVE_TYPE >::source_location;
```
The downside is having this extra stuff to remember to put in kernels. Maybe there's a better way to associate kernel class with source file name?

These ideas aren't too pretty, just thought I'd mention them. They allow us to gets rid of `JITTI_DECL`/`JITTI_TPARAM` macros entirely and unify the two dispatchers, possibly even making them specializations as the comment above suggests, and allowing for a per-kernel JITting decision.

Reply:
I think the end goal is to always use `JITTI`, although that doesn't mean that we need to always jit things at run-time. We need to have the capability to pre-jit everything at build-time, and I think that would suffice.